rust/headless-client/src/ipc_service.rs

```diff
@@ -133,24 +116,48 @@ fn run_smoke_test() -> Result<()> {
     crate::setup_stdout_logging()?;
     let rt = tokio::runtime::Runtime::new()?;
     let _guard = rt.enter();
+    let mut signals = signals::Terminate::new()?;
 
     // Couldn't get the loop to work here yet, so SIGHUP is not implemented
     rt.block_on(async {
         device_id::get_or_create().context("Failed to read / create device ID")?;
         let mut server = IpcServer::new(ServiceId::Prod).await?;
-        Handler::new(&mut server).await?.run().await;
+        let _ = Handler::new(&mut server).await?.run(&mut signals).await;
         Ok::<_, anyhow::Error>(())
     })
 }
 
-async fn ipc_listen() -> Result<std::convert::Infallible> {
+/// Run the IPC service and terminate gracefully if we catch a terminate signal
+///
+/// If an IPC client is connected when we catch a terminate signal, we send the
+/// client a hint about that before we exit.
+async fn ipc_listen(signals: &mut signals::Terminate) -> Result<()> {
     // Create the device ID and IPC service config dir if needed
     // This also gives the GUI a safe place to put the log filter config
     device_id::get_or_create().context("Failed to read / create device ID")?;
     let mut server = IpcServer::new(ServiceId::Prod).await?;
     loop {
-        Handler::new(&mut server).await?.run().await;
+        let mut handler_fut = pin!(Handler::new(&mut server));
+        let Some(handler) = poll_fn(|cx| {
```

Comment:
A lot of nesting going on here. Could this be simpler if we used a `select()` instead and matched on the `Either`?